### ISSUE TITLE
Gendered emoji render as two characters

### ISSUE THREAD

**Issue by james-howard** [codemirror/CodeMirror#4488]
Environment: Safari 10.0.2, Chrome 55.0.2883.95 on macOS 10.2.2 (may reproduce elsewhere, but this is what I've tested).

Steps to reproduce:
1. In codemirror, press ctrl-cmd-space to bring up the emoji picker
2. Choose an emoji that has gender or skin tone options (I used the police woman).
3. Insert the emoji

Results:
Emoji renders as two characters:
<img width="616" alt="screen shot 2017-01-09 at 1 03 59 pm" src="https://cloud.githubusercontent.com/assets/2006254/21783431/20e76616-d66c-11e6-9bfb-75ef2751d0ac.png">

Expected results:
Emoji renders as a single character:
<img width="28" alt="screen shot 2017-01-09 at 1 05 15 pm" src="https://cloud.githubusercontent.com/assets/2006254/21783477/501125bc-d66c-11e6-8645-7d263de0a2a5.png">

Notes:
I expect that this is due to the way these characters are encoded.

I investigate this and provide a pull request to fix the issue if possible.

**Comment by marijnh** [comment]
This is probably happening because CodeMirror turns off ligatures in its CSS, because those cause problems with cursor positioning.

**Comment by james-howard** [comment]
Thanks for the pointer. As a quick test, I disabled that CSS and can still reproduce the issue, so I suspect it's something else.

**Comment by james-howard** [comment]
Some more investigation indicates that the issue is more complex than I initially imagined. These emoji are actually multiple emojis joined with a zero width joiner, u200d. u200d is treated as a special character by CodeMirror, so that explains why they are getting split into multiple parts. 

However, even if we don't treat this as a special character, it's still not enough as CodeMirror is using string length in a number of places as a proxy for editable characters, which is unfortunately a fallacy (I could be using the wrong terminology, by editable characters I mean characters that can be navigated amongst as atomic units). Even doing something like [...string].length isn't correct in the face of these zero width joined emojis, in terms of figuring out how many editable characters are in the document.

I'm probably not going to try anymore to write a patch for this because it would be a very big change and I suspect it could have a pretty negative impact on performance even if I got it right. I've also learned that outside of Xcode, no code editors support these emojis properly, so at the least CodeMirror is in good company.

**Comment by marijnh** [comment]
> as CodeMirror is using string length in a number of places as a proxy for editable characters

Can you give examples of such places?

**Comment by james-howard** [comment]
Well, at a minimum, there's a couple of places in `buildToken` in line/line-data.js. There may be others. I think to get it right, you'd have to get some facility for doing proper grapheme iteration in JS, and then use it religiously rather than ever using string length or iterating over characters or slicing strings directly. A lot of work, I think.

**Comment by marijnh** [comment]
> Well, at a minimum, there's a couple of places in buildToken in line/line-data.js.

Those appear to be using string length to track offset into a string, which is entirely okay. The line that tracks the column is indeed suspicious, but that'll only cause tab characters to have the wrong with in selected circumstances, so that's a minor issue.

If you find a reliable and fast way to query font ligatures from JavaScript, I'm interested. But until then, slightly shaky workarounds are what we have to live with.

**Comment by philippkuehn** [comment]
@marijnh Emojis have become very important nowadays. Is there a chance that this can be solved? I just see that this problem still exists in CodeMirror 6.

**Comment by marijnh** [comment]
> Is there a chance that this can be solved?

I don't know. Let me know when you find a robust and performant way to figure out which characters are going to be joined into a single glyph (which is font-dependent).

**Comment by philippkuehn** [comment]
Damn 😅 Do you know how other editors like monaco handles this problem?

**Comment by marijnh** [comment]
I haven't looked.

I guess we could do a crude approximation, just assuming that zero-width joiners always group multiple characters into a glyph. It'll look a bit weird when your font _doesn't_ support the ligature, but might still be better than the current behavior. I'll experiment with that in CodeMirror 6.

Attached patch removes special handling of zero-width joiners, to at least not actively sabotage the rendering of joined emoji.

**Comment by philippkuehn** [comment]
Yeah, rendering is better now but backspace doesn't work as expected 😅

![May-26-2020 15-06-11](https://user-images.githubusercontent.com/2500670/82904232-73679100-9f62-11ea-88ef-3a0fee9a2e84.gif)

Since this is probably a bigger problem, I try to write my own `backspace` command, and recognize emojis with `grapheme-splitter` https://github.com/orling/grapheme-splitter.

**Comment by marijnh** [comment]
>  I try to write my own backspace command

(You'll want to do the same for horizontal arrow motion.)

**Comment by philippkuehn** [comment]
Good catch. Let’s see how far I can get.
